convex/lib/rowLevelSecurity.ts

```diff
@@ -0,0 +1,404 @@
+    paginationOpts: PaginationOptions
+  ): Promise<PaginationResult<DocumentByInfo<T>>> {
+    return this.fullTableScan().paginate(paginationOpts);
+  }
+  collect(): Promise<DocumentByInfo<T>[]> {
+    return this.fullTableScan().collect();
+  }
+  take(n: number): Promise<DocumentByInfo<T>[]> {
+    return this.fullTableScan().take(n);
+  }
+  first(): Promise<DocumentByInfo<T> | null> {
+    return this.fullTableScan().first();
+  }
+  unique(): Promise<DocumentByInfo<T> | null> {
+    return this.fullTableScan().unique();
+  }
+  [Symbol.asyncIterator](): AsyncIterator<DocumentByInfo<T>, any, undefined> {
+    return this.fullTableScan()[Symbol.asyncIterator]();
+  }
+}
+
+class WrapReader<Ctx, DataModel extends GenericDataModel>
+  implements DatabaseReader<DataModel>
+{
+  ctx: Ctx;
+  db: DatabaseReader<DataModel>;
+  readAccessRules: Rules<Ctx, DataModel>;
+
+  constructor(
+    ctx: Ctx,
+    db: DatabaseReader<DataModel>,
+    readAccessRules: Rules<Ctx, DataModel>
+  ) {
+    this.ctx = ctx;
+    this.db = db;
+    this.readAccessRules = readAccessRules;
+  }
+
+  async predicate<T extends GenericTableInfo>(
+    tableName: string,
+    doc: DocumentByInfo<T>
+  ): Promise<boolean> {
+    if (!(tableName in this.readAccessRules)) {
+      return true;
+    }
+    return await this.readAccessRules[tableName]!(this.ctx, doc);
+  }
+
+  async get<TableName extends string>(id: GenericId<TableName>): Promise<any> {
+    const doc = await this.db.get(id);
+    if (doc && (await this.predicate(id.tableName, doc))) {
+      return doc;
+    }
+    return null;
+  }
+
+  query<TableName extends string>(
+    tableName: TableName
+  ): QueryInitializer<NamedTableInfo<DataModel, TableName>> {
+    return new WrapQueryInitializer(
+      this.db.query(tableName),
+      async (d) => await this.predicate(tableName, d)
+    );
+  }
+}
+
+class WrapWriter<Ctx, DataModel extends GenericDataModel>
+  implements DatabaseWriter<DataModel>
+{
+  ctx: Ctx;
+  db: DatabaseWriter<DataModel>;
+  reader: DatabaseReader<DataModel>;
+  writeAccessRules: Rules<Ctx, DataModel>;
+
+  async predicate<T extends GenericTableInfo>(
+    tableName: string,
+    doc: DocumentByInfo<T>
+  ): Promise<boolean> {
+    if (!(tableName in this.writeAccessRules)) {
+      return true;
+    }
+    return await this.writeAccessRules[tableName]!(this.ctx, doc);
+  }
+
+  constructor(
+    ctx: Ctx,
+    db: DatabaseWriter<DataModel>,
+    readAccessRules: Rules<Ctx, DataModel>,
+    writeAccessRules: Rules<Ctx, DataModel>
+  ) {
+    this.ctx = ctx;
+    this.db = db;
+    this.reader = new WrapReader(ctx, db, readAccessRules);
+    this.writeAccessRules = writeAccessRules;
+  }
+  async insert<TableName extends string>(
+    table: TableName,
+    value: any
+  ): Promise<any> {
+    // No auth check on insert.
+    return await this.db.insert(table, value);
+  }
+  async checkAuth<TableName extends string>(id: GenericId<TableName>) {
+    const doc = await this.get(id);
```

Comment:
Let's also make it clear that all operations will now do reads too - so a batch delete function should definitely try to parallelize, etc! It's actually so nice that our writer API was already async - this wouldn't be possible with the synchronous writing API
I've in the past done `patch` instead of get & replace b/c I thought it'd be faster. In this world, it might be worth mentioning that it won't be, you may as well grab & replace.

Reply:
i don't think there's much of a difference here. left a comment to that effect